### ISSUE TITLE
Native messenger doesn't seem to be able to open vimr?

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [cmcaine/tridactyl#684]
I've set `editorcmd` to `vimr --wait`, but now `<Ctrl-i>` has no effect whatsoever. (Notably, VimR *doesn't* launch.)

1. Is there an easy way to revert a setting to its default, other than looking it up in the source, yet?
2. Maybe add VimR to the default lookup system — `"auto"` opens an ancient MacVim, on my system. `(=`

**Comment by glacambre** [comment]
`unset editorcmd` should do the trick.

Have you tried `vimr --wait --nvim`? I'm don't really know what it does but it looks like it might be needed: https://github.com/qvacua/vimr/wiki

**Comment by ELLIOTTCABLE** [comment]
`--nvim` is for passing unrelated commands to the underlying NeoVim server.

Thing is, the VimR GUI application doesn't even launch, at all — even `:set editorcmd vimr`, non-wait mode, doesn't launch and divest a child-process as expected.

Is there a way I can debug the native-messenger from the command-line? Or see some output in Tridactyl itself?

**Comment by glacambre** [comment]
Oh, I remember an issue about things being weird in OSX: https://github.com/cmcaine/tridactyl/issues/451

In order to debug the native messenger you could try using `:set logging.native debug` and then open the console in the current page (ctrl+shift+I) and the one for Tridactyl that you'll find in `about:debugging` and see if there's anything interesting in the logs.

You could also try manually running the native messenger from your terminal like this:
`printf '%c\0\0\0{"cmd": "run", "command": "vimr --wait"}' 40 | ~/.local/share/tridactyl/native_main.py`
If you want to change the command that is run, make sure to replace "40" with the number of characters of the json object.

**Comment by bovine3dom** [comment]
@ELLIOTTCABLE stupid question: does typing `vimr` in a terminal actually open it? What about `! vimr` in Tridactyl?

If the former doesn't work, then that's just OSX being a bit pants and you should update MacVim and use that instead.

If only the latter doesn't work, then vimr is somehow adding itself to your path in your terminal, so you should just use the absolute path to the binary.

**Comment by bovine3dom** [comment]
Also, are you 100% sure you did

`It can be installed via Preferences ➝ General ➝ "Copy 'vimr' CLI tool..." button:` in vimr?

**Comment by ELLIOTTCABLE** [comment]
Ahhah, it *does* seem to be related to `$PATH`.

1. `vimr` in my shell works plenty well
2. `:! vimr` in Tridactyl has no effect
3. `:! /usr/local/bin/vimr` in Tridactyl works great (yay!)

So, how is Tridactyl evaluating paths? Presumably it's the `$PATH` as seen by the native-messenger binary? How is that launched/invoked — as a child-process of FireFox? If so, then it wouldn't have *any* user's `$PATH` setup … but I can't be the first one to have run into this, if so?

**Comment by ELLIOTTCABLE** [comment]
It looks like this might be useful background information:

[Why can't I set PATH with EnvPane?](https://github.com/hschmidt/EnvPane#why-cant-I-set-path-with-envpane)

> That's because OS X treats PATH differently to other environment variables, I suspect for security reasons. The special treatment differs from version to version of OS X but in a nutshell there are two issues: Firstly, launchd will forcefully set PATH to a fixed value, …

tl;dr it's a mess. Maybe add a note to the README that `$PATH` is not available / replicated, and to use absolute paths if possible?

**Comment by protrolium** [comment]
would these $PATH issues have something to with why I am unable to install native messenger when running the `curl` command in macOS 10.13.6 terminal ? 

```
Error: Python 3 must exist in PATH.\e[0m
Please install it and run this script again.\e[0m
```

**Comment by bovine3dom** [comment]
This might sound like a silly question, but is Python 3 on your path? What happens when you run `python3` in your terminal?

**Comment by protrolium** [comment]
Not at all. Looks like I had not set python3.x.x with global command within pyenv. native messenger now installed. Thanks!

**Comment by Shaddy** [comment]
Hi all,

I am having a similar problem as @ELLIOTTCABLE  had.

If I do `! echo $PATH` this is what I see:

`/usr/bin:/bin:/usr/sbin:/sbin`

And the vimr in OSX (unless I disable the integrity) is in `/usr/local/bin`. Is it possible to specify somehow the full path?

Using `set editorcmd /usr/local/bin/vimr` does not work.

**Comment by Shaddy** [comment]
Btw, if I run `! /usr/local/bin/vimr` it works w/o problems :(

**Comment by Shaddy** [comment]
Damn I just realized it was working, never test by using `editor` alone...

**Comment by Salomanuel** [comment]
I'm new to this amazing plugin and I had some issues making this work, but this thread immensely helped me :)

all this made the trick:

1) install vimr, even better if with brew
2) from vimr, install copy vimr CLI tool: 
    It can be installed via Preferences ➝ General ➝ "Copy 'vimr' CLI tool..." button: in vimr
3) after clicking that button, put it in /usr/local/bin
4) back to firefox, `:set editorcmd vimr --wait`

this worked for me




BUT NOW I'm facing another issue:

1) from insert mode inside a text field, I press <Ctrl-i>
2) vimr opens and I do my business
3) I save with `:w`

4a) I quit with <CMD-q>
5a) I go back to edit mode, everything is fine, it's a nice day

4b) I quit with `:q`
5b) I am back on the firefox page, in insert mode, but I'm not in the text field anymore and all the keyboard commands don't work anymore. Not even <ESC> works and every keypress just makes a flash and a sound. 
6b) If I click somewhere with the mouse, everything works as before. But, as we all know, the mouse is downstairs, that's why we installed this addon.

Any idea on how to tackle this?

**Comment by makovich** [comment]
For those who are going to dig in later.

1. My first attack angle was [poUsePath](https://github.com/nim-lang/Nim/blob/4459c272ecb8ddb964fa2d34a4dcef3368d807b5/lib/pure/osproc.nim#L44) flag in `native_main`. There is a fun [comment](https://github.com/nim-lang/Nim/blob/4459c272ecb8ddb964fa2d34a4dcef3368d807b5/lib/pure/osproc.nim#L1175) regarding MacOS. But... turns out that it's not the case. Nothing changed after rebuild.

3. `tri.native.run('env')` output in browser's console gave me this:
    ```shell
    PATH=/usr/bin:/bin:/usr/sbin:/sbin
    SHELL=/bin/zsh
    MOZ_APP_SILENT_START=
    TMPDIR=/var/folders/6c/045qx8w55_g_xdkpjjvwgkcc0000gn/T/
    Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.0FuoZuYqfE/Render
    MOZ_CRASHREPORTER_PING_DIRECTORY=/Users/_/Library/Application Support/Firefox/Pending Pings
    MallocNanoZone=0
    XRE_RESTARTED_BY_PROFILE_MANAGER=
    SYSTEM_VERSION_COMPAT=0
    XRE_PROFILE_LOCAL_PATH=
    USER=_
    XRE_START_OFFLINE=
    SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.Kav5K6san6/Listeners
    __CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0
    PWD=/Users/_/.local/share/tridactyl
    XRE_PROFILE_PATH=
    TZ=UTC
    XPC_FLAGS=0x0
    XPC_SERVICE_NAME=org.mozilla.firefox.12456
    SHLVL=1
    HOME=/Users/_
    MOZ_CRASHREPORTER_EVENTS_DIRECTORY=/Users/_/HajhHfx.ff/crashes/events
    LOGNAME=morse
    MOZ_LAUNCHED_CHILD=
    MOZ_CRASHREPORTER_RESTART_ARG_1=-foreground
    XRE_BINARY_PATH=
    MOZ_CRASHREPORTER_RESTART_ARG_0=/Applications/Firefox.app/Contents/MacOS/firefox
    DISPLAY=/private/tmp/com.apple.launchd.0KgLQXIRCw/org.macosforge.xquartz:0
    MOZ_CRASHREPORTER_RESTART_ARG_2=
    XUL_APP_FILE=
    MOZ_CRASHREPORTER_DATA_DIRECTORY=/Users/_/Library/Application Support/Firefox/Crash Reports
    MOZ_CRASHREPORTER_STRINGS_OVERRIDE=/Applications/Firefox.app/Contents/Resources/browser/crashreporter-override.ini
    _=/usr/bin/env
    ```
    This makes impression that Firefox fork/exec into kinda `native_main` and so sets `PATH` by itself.  Entries in `PATH` here is not something that I can find myself somewhere in `/etc` or near.

4. Workaround for things like `mpvsafe` would be `PATH=/path/to/another/bin:$PATH`:
    ```js
    set exaliases.mpv js -p tri.excmds.shellescape(JS_ARG).then(url => tri.excmds.exclaim_quiet('PATH=$HOME/bin:$PATH mpv --no-terminal ' + url))
    ```

**Comment by vamega** [comment]
So my owned crazed approach to this was to do the following, since I saw in @makovich's output that the SHELL is set correctly.

```javascript
tri.excmds.exclaim(`$SHELL --login -c "~/.config/tridactyl/scripts/script.sh"`)
```

For `ZSH` at the very least this will source variables in `~/.zshenv`. You can also pass `--interactive` if you'd like it to source `~/.zshrc`.

That pattern now let's me write scripts a lot more easily, since using full paths introduces other complexities in my setup.

**Comment by Reefersleep** [comment]
> I'm new to this amazing plugin and I had some issues making this work, but this thread immensely helped me :)
> 
> all this made the trick:
> 
>     1. install vimr, even better if with brew
> 
>     2. from vimr, install copy vimr CLI tool:
>        It can be installed via Preferences ➝ General ➝ "Copy 'vimr' CLI tool..." button: in vimr
> 
>     3. after clicking that button, put it in /usr/local/bin
> 
>     4. back to firefox, `:set editorcmd vimr --wait`
> 
> 
> this worked for me
> 
> BUT NOW I'm facing another issue:
> 
>     1. from insert mode inside a text field, I press `<Ctrl-i>`
> 
>     2. vimr opens and I do my business
> 
>     3. I save with `:w`
> 
> 
> 4a) I quit with `<CMD-q>` 5a) I go back to edit mode, everything is fine, it's a nice sunny day
> 
> 4b) I quit with `:q` 5b) I am back on the firefox page, in insert mode, but I'm not in the text field anymore and all the keyboard commands don't work anymore. Not even `<ESC>` works and every keypress just makes a flash and a sound. 6b) If I click somewhere with the mouse, everything works as before. But, as we all know, the mouse is downstairs, that's why we installed this addon.
> 
> Any idea on how to tackle this?

I had and fixed a similarish issue, but using Macvim. 
I installed Macvim with `brew install macvim`.
I have 
```
:set editorcmd /opt/homebrew/bin/mvim -f
```
which works great with <Ctrl-i> for _opening_ Macvim and loading the text in the browser input field in the buffer, but on `:wq` or `ZZ`, the edited buffer would not be pasted back into the browser input field. This would only happen on `<Cmd-q>` or otherwise closing the Macvim application, which keeps running after `:wq`/`ZZ` - according to the default settings. However, since I don't use Macvim for anything else, I could change this option in Macvim:
`Preferences -> General -> After last window closes: Quit Macvim` 

And now, `:wq`/`ZZ` works as intended, hoorray!
